### PR TITLE
HPCC-32031 Generate summary information in workunit to speed up file list operations

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: common/workunit/workunit.cpp [hunk 1]
```diff
@@ -8721,6 +8725,46 @@ void CLocalWorkUnit::setDebugValue(const char *propname, const char *value, bool
...
+};
+
+bool CLocalWorkUnit::getSummary(SummaryType type, SummaryMap &map) const
+{
+    VStringBuffer xpath("Summaries/%s", summaryTypeName(type));
+    CriticalBlock block(crit);
+    IPropertyTree *list = p->queryPropTree(xpath);
+    if (!list)
+        return false;
+    Owned<IPropertyTreeIterator> r = list->getElements("v");
```
Comment (ghalliday) [inline common/workunit/workunit.cpp:8750]: I wonder if it would be much better to store all the (key,value) pairs as a single serialized string value.  

It is likely to be more efficient to store and retrieve.  @jakesmith how do you think the performance/storage requirements would compare.

File: ecl/hqlcpp/hqlres.cpp [hunk 2]
```diff
@@ -648,7 +648,8 @@ bool ResourceManager::flush(StringBuffer &filename, const char *basename, bool f
         }
         fwrite(s.data.get(), 1, s.data.length(), bin);
         fclose(bin);
-        fprintf(f, " .size %s,%u\n", label.str(), (unsigned)s.data.length());
+        if (!generateClang)
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlres.cpp:651]: this looks like an unrelated bug.  If so it will need to target an earlier build.

**Review by ghalliday — COMMENTED** [review]
@richardkchapman looks good.

File: ecl/hqlcpp/hqlres.cpp (hunk 2 above)
Comment (richardkchapman) [reply ecl/hqlcpp/hqlres.cpp:651]: I don't know if it's in general a change that is needed - but I needed it to be able to compile ANY workunit on my current mac system

**Comment by richardkchapman** [comment]
Couple of things to think about:
1. Should we use a more compact format to store? Less readable, but potentially (slightly?) more efficient.
2. Should we be using case-sensitive maps and/or lower-casing filenames before inserting them?

File: common/workunit/workunit.cpp [hunk 3]
```diff
@@ -8721,6 +8725,53 @@ void CLocalWorkUnit::setDebugValue(const char *propname, const char *value, bool
...
+    ForEach(*v)
+    {
+        const char *name = v->query().queryProp(".");
+        map[name] = false;
+    }
+    Owned<IPropertyTreeIterator> o = list->getElements("o");
+    ForEach(*o)
+    {
+        const char *name = o->query().queryProp(".");
+        if (map.find(name) != map.end())
```
Comment (richardkchapman) [inline common/workunit/workunit.cpp:8760]: Should say == map.end() I think?

**Comment by richardkchapman** [comment]
@ghallidat rebased to 9.8.x and marked as ready. I have made the stored format more compact, and made teh table case-insensitive

**Comment by ghalliday** [comment]
From regression test:
aggsq2,alljoin - spill:1 not classified as a spill
alien: jobtemp not classified as a spill (probably should be.  They could be a separate category of spills)
apersistschedule: is it worth separating persists into a separate spill category?

**Comment by ghalliday** [comment]
@jakesmith is it worth recording spills? They should be ignored for roxie targets (that can be in a later PR), but is there any benefit in having a summary list for hthor/thor jobs?

File: ecl/hqlcpp/hqlcpp.ipp [hunk 4]
```diff
@@ -2140,6 +2141,25 @@ protected:
...
+            SummaryMap &map = summaries[(int) type];
+            SummaryFlags flags = SummaryFlags::None;
+            if (isOpt)
+                flags |= SummaryFlags::IsOpt;
+            if (isSigned)
+                flags |= SummaryFlags::IsSigned;
+            if (map.find(name) == map.end())
+                map[name] = flags;
+            else
+                map[name] = map[name] & flags;
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.ipp:2160]: can avoid re-searching.  I will release an improvement.

**Review by ghalliday — APPROVED** [review]

File: common/pkgfiles/referencedfilelist.cpp [hunk 5]
```diff
@@ -1018,6 +999,62 @@ bool ReferencedFileList::addFilesFromQuery(IConstWorkUnit *cw, const IHpccPackag
                 ensureFile(logicalName, flags, NULL, false, &subfileNames);
         }
     }
+    else
+    {
+        Owned<IConstWUGraphIterator> graphs = &cw->getGraphs(GraphTypeActivities);
+        ForEach(*graphs)
+        {
+            Owned <IPropertyTree> xgmml = graphs->query().getXGMMLTree(false, false);
+            Owned<IPropertyTreeIterator> iter = xgmml->getElements("//node[att/@name='_*ileName']");
```
Comment (jakesmith) [inline common/pkgfiles/referencedfilelist.cpp:1008]: '_*ileName' looks odd, is it because it can be _fileName, or spillFileName or similar? could do with a comment.

Comment (jakesmith) [reply common/pkgfiles/referencedfilelist.cpp:1008]: (but I now see this is just moved code, not new code)

File: common/workunit/workunit.cpp [hunk 6]
```diff
@@ -8721,6 +8725,65 @@ void CLocalWorkUnit::setDebugValue(const char *propname, const char *value, bool
...
+{
+    switch (type)
+    {
+    case SummaryType::ReadFile: return "ReadFile";
+    case SummaryType::ReadIndex: return "ReadIndex";
+    case SummaryType::WriteFile: return "WriteFile";
+    case SummaryType::WriteIndex: return "WriteIndex";
+    case SummaryType::PersistFile: return "PersistFile";
+    case SummaryType::SpillFile: return "SpillFile";
+    case SummaryType::JobTemp: return "JobTemp";
```
Comment (jakesmith) [inline common/workunit/workunit.cpp:8738]: I am not sure of the usefulness/intended purposes of recording SpillTemp/JobTemp files

File: common/workunit/workunit.cpp [hunk 7]
```diff
@@ -8721,6 +8725,65 @@ void CLocalWorkUnit::setDebugValue(const char *propname, const char *value, bool
...
+        {
+            char *end = nullptr;
+            SummaryFlags flags = (SummaryFlags) strtol(name, &end, 16);
+            if (*end!=':')
+                return false; // unrecognized format
+            name = end+1;
+            if (map.find(name) == map.end())
+                map[name] = flags;
+            else
+                map[name] = map[name] & flags;
```
Comment (jakesmith) [inline common/workunit/workunit.cpp:8767]: trivial: could be:
```
            auto it = map.find(name);
            if (it == map.end())
                map[name] = flags;
            else
                it->second &= flags;
```

**Review by jakesmith — APPROVED** [review]
@richardkchapman - looks good to me, couple of trivial comments (1 question re. SpillFile/JobTemp's)

File: common/workunit/workunit.cpp (hunk 7 above)
Comment (ghalliday) [reply common/workunit/workunit.cpp:8767]: yes, I will fix in a follow on PR (along with the similar code below)
